pontoon/base/views.py

```diff
@@ -348,6 +348,87 @@ def get_translation_history(request):
     return JsonResponse(payload, safe=False)
 
 
+@utils.require_AJAX
+@login_required(redirect_field_name='', login_url='/403')
+@transaction.atomic
+def approve_translation(request):
+    """Approve given translation."""
+    try:
+        t = request.POST['translation']
+        paths = request.POST.getlist('paths[]')
+    except MultiValueDictKeyError as e:
+        return HttpResponseBadRequest('Bad Request: {error}'.format(error=e))
+
+    translation = get_object_or_404(Translation, pk=t)
+    project = translation.entity.resource.project
+    locale = translation.locale
+    user = request.user
+
+    # Read-only translations cannot be approved
+    if utils.readonly_exists(project, locale):
+        return HttpResponseForbidden(
+            "Forbidden: This translation is in read-only mode"
+        )
+
+    if translation.approved:
+        return HttpResponseForbidden(
+            "Forbidden: This translation is already approved"
+        )
+
+    # Only privileged users can approve translations
+    if not user.can_translate(locale, project):
+        return HttpResponseForbidden(
+            "Forbidden: You don't have permission to approve this translation."
+        )
+
+    # Check for errors.
+    # Checks are disabled for the tutorial.
+    use_checks = project.slug != 'tutorial'
+
+    if use_checks and translation.errors.exists():
+        return JsonResponse({
+            'failedChecks': translation.errors.all(),
+        })
```

Comment:
We should return all failed checks here, including warnings. And they should be formatted in the same fashion as `run_checks` formats them. 

You can also take care of this in followup bug that will implement checks, in which case you might want to error here, so we don't forget about this use case (and only focus on saving translations by mistake). Seems like we don't have a bug on file for implementing checks in Translate.Next, yet?

Reply:
OK, filed https://bugzilla.mozilla.org/show_bug.cgi?id=1518820 then, and I'll remove this to just raise an error.